Source/Lib/Codec/EbEncDecProcess.c

```diff
@@ -1730,9 +1718,21 @@ static void ReconOutput(
         outputReconPtr->pts = pictureControlSetPtr->pictureNumber;
     }
 
+    // The totalNumberOfReconFrames counter has to be write/read protected as
+    //   it is used to determine the end of the stream.  If it is not protected
+    //   the encoder might not properly terminate.
+    EbBlockOnMutex(encodeContextPtr->terminatingConditionsMutex);
+
+    // START READ/WRITE PROTECTED SECTION
+    if (encodeContextPtr->totalNumberOfReconFrames == encodeContextPtr->terminatingPictureNumber) {
+        outputReconPtr->nFlags = EB_BUFFERFLAG_EOS;
+    }
+    encodeContextPtr->totalNumberOfReconFrames++;
+
     // Post the Recon object
```

Comment:
Are EbGetEmptyObject and  EbPostFullObject need to be protected by terminatingConditionsMutex?
If not, only protect totalNumberOfReconFrames is enough.